dlgr/griduniverse/static/scripts/items.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * Representation of a game item, which for the moment is limited to a
+ * simple Food type.
+ */
+
+export class Item {
+  constructor(id, itemId, maturity, remainingUses) {
+    this.id = id;
+    this.itemId = itemId;
+    this.maturity = maturity;
+    this.remainingUses = remainingUses;
+
+    // XXX Maybe we can avoid this copy of every shared value
+    // to every instance, but going with it for now.
+    Object.assign(this, settings.item_config[this.itemId]);
+  }
+
+  /**
+   * Calculate a color based on sprite definition and maturity
+   */
+  get color() {
+    let immature, mature;
+
+    if (this.sprite.includes(",")) {
+      [immature, mature] = this.sprite.split(",");
+      // For now, assume these are hex colors
+    } else {
+      immature = mature = this.sprite;
+    }
+
+    return rgbOnScale(
+      hexToRgbPercentages(immature),
+      hexToRgbPercentages(mature),
+      this.maturity
+    );
+  }
+}
+
+/**
+ * Manages Items that sit on the grid.
+ *
+ * Items currently being carried by a Player are not included.
+ * These are instead held by reference in the Play instances,
+ * (named `current_item`).
+ */
+export class GridItems {
+  constructor() {
+    this._itemsByPosition = new Map();
+    this._positionsById = new Map();
+  }
+
+  add(item, position) {
+    this._itemsByPosition.set(JSON.stringify(position), item);
+    this._positionsById.set(item.id, JSON.stringify(position));
+  }
+
+  atPosition(position) {
+    const key = JSON.stringify(position);
+    return this._itemsByPosition.get(key);
```

Comment:
I think this should explicitly return a `null` (e.g. ` || null`) if nothing is found, if only because `undefined` seems like kind of a weird value for this to return.